### ISSUE TITLE
query_by_weights inconsistently affects ranking even for equally weighted fields

### ISSUE THREAD

**Issue by superkelvint** [typesense/typesense#440]
## Description
Adding a query_by_weight parameter of "1,1" for a 2-field search causes ranking to change which appears to violate the documentation. 

## Steps to reproduce

See the following test case in Java.

`
import org.typesense.api.Client;
import org.typesense.api.Configuration;
import org.typesense.api.Documents;
import org.typesense.model.*;
import org.typesense.resources.Node;

import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class TestQueryByWeights {
    public static void main(String[] args) {
        Node node = new Node(
                "http",       // For Typesense Cloud use https
                "localhost",  // For Typesense Cloud use xxx.a1.typesense.net
                "8108"        // For Typesense Cloud use 443
        );

        ArrayList<Node> nodes = new ArrayList<>();
        nodes.add(node);

        Configuration configuration = new Configuration(nodes, Duration.ofSeconds(2),"5gT4XrHnWVIeUTa7cx9ns1QMUCkj4HMBs69hOIMljGavZtLs");

        Client client = new Client(configuration);

        ArrayList<Field> fields = new ArrayList<>();
        fields.add(new Field().name("fieldA").type(Field.TypeEnum.STRING));
        fields.add(new Field().name("fieldB").type(Field.TypeEnum.STRING));
        String collection = "testQueryByWeights";
        client.collections(collection).delete();
        CollectionSchema collectionSchema = new CollectionSchema();
        collectionSchema.name(collection).fields(fields);
        client.collections().create(collectionSchema);

        HashMap doc1 = new HashMap();
        doc1.put("id", "doc1");
        doc1.put("fieldA", "foo same");
        doc1.put("fieldB", "bar same");

        HashMap doc2 = new HashMap();
        doc2.put("id", "doc2");
        doc2.put("fieldA", "bar same");
        doc2.put("fieldB", "foo same");

        Documents documents = client.collections(collection).documents();
        documents.create(doc1);
        documents.create(doc2);

        SearchParameters searchParameters = new SearchParameters()
                .q("foo same").addQueryByItem("fieldA,fieldB").addPrefixItem(true)
                ;
        SearchResult searchResult = client.collections(collection).documents().search(searchParameters);
        List<SearchResultHit> hits = searchResult.getHits();
        assertEquals(hits.get(0).getDocument().get("id"), "doc1");
        assertEquals(hits.get(1).getDocument().get("id"), "doc2");

        searchParameters = new SearchParameters()
                .q("foo same").addQueryByItem("fieldA,fieldB").addQueryByWeightsItem("1,1").addPrefixItem(true)
        ;
        searchResult = client.collections(collection).documents().search(searchParameters);
        hits = searchResult.getHits();
        //==============================================================================================
        // The following asserts fail
        //==============================================================================================
        //        assertEquals(hits.get(0).getDocument().get("id"), "doc1");
        //        assertEquals(hits.get(1).getDocument().get("id"), "doc2");

        searchParameters = new SearchParameters()
                .q("foo same").addQueryByItem("fieldB,fieldA").addPrefixItem(true)
        ;
        searchResult = client.collections(collection).documents().search(searchParameters);
        hits = searchResult.getHits();
        assertEquals(hits.get(0).getDocument().get("id"), "doc2");
        assertEquals(hits.get(1).getDocument().get("id"), "doc1");

        searchParameters = new SearchParameters()
                .q("foo same").addQueryByItem("fieldB,fieldA").addQueryByWeightsItem("1,1").addPrefixItem(true)
        ;
        searchResult = client.collections(collection).documents().search(searchParameters);
        hits = searchResult.getHits();
        //==============================================================================================
        // The following asserts, however, pass.
        //==============================================================================================
        assertEquals(hits.get(0).getDocument().get("id"), "doc2");
        assertEquals(hits.get(1).getDocument().get("id"), "doc1");

        searchParameters = new SearchParameters()
                .q("foo same").addQueryByItem("fieldB,fieldA").addQueryByWeightsItem("1,2").addPrefixItem(true)
        ;
        searchResult = client.collections(collection).documents().search(searchParameters);
        hits = searchResult.getHits();
        // these asserts pass
        assertEquals(hits.get(0).getDocument().get("id"), "doc1");
        assertEquals(hits.get(1).getDocument().get("id"), "doc2");

    }

    private static void assertEquals(Object expected, Object actual) {
        if (!expected.equals(actual)) throw new AssertionError("Expected " + expected + " got " + actual);
    }
}

`

## Expected Behavior

For equally weighted fields, having a query_by_weights should *not* affect ranking. 

## Actual Behavior

Ranking if affected.

## Metadata

**Typesense Version**: 0.21.0

**OS**: Ubuntu 20.04

**Comment by kishorenc** [comment]
@superkelvint 

We've improved field weighting greatly in the upcoming 0.22 release, for which we have RC builds. 

Can you please try against [0.22.rcs37](https://dl.typesense.org/releases/0.22.0.rcs37/typesense-server-0.22.0.rcs37-linux-amd64.tar.gz)?

**Comment by superkelvint** [comment]
Hi @kishorenc, I just tried against 0.22.rcs37 and the test still fails. I can provide the unit test in a different language if you can't run the Java class easily.

**Comment by superkelvint** [comment]
@kishorenc I found something else that might indicate you're using a signed byte to represent query_by_weight values because for weights > 127, they seem to be ignored or inverted. There's no indication in the documentation that indicates an upper bound for query weights. Let me know if you'd like me to open a separate issue to track this. 

This is happening on 0.22.rcs37 btw. 

Here's a test case (again in Java):

`
HashMap doc1 = new HashMap();
        doc1.put("id", "doc1");
        doc1.put("fieldA", "foo same");
        doc1.put("fieldB", "bar same");

        HashMap doc2 = new HashMap();
        doc2.put("id", "doc2");
        doc2.put("fieldA", "bar same");
        doc2.put("fieldB", "foo same");

        Documents documents = client.collections(collection).documents();
        documents.create(doc1);
        documents.create(doc2);

        SearchParameters searchParameters = new SearchParameters()
                .q("foo same").addQueryByItem("fieldA,fieldB").addPrefixItem(true)
                ;
        SearchResult searchResult = client.collections(collection).documents().search(searchParameters);
        List<SearchResultHit> hits = searchResult.getHits();
        assertEquals(hits.get(0).getDocument().get("id"), "doc1");
        assertEquals(hits.get(1).getDocument().get("id"), "doc2");

        searchParameters = new SearchParameters()
                .q("foo same").addQueryByItem("fieldA,fieldB").addQueryByWeightsItem("1,2").addPrefixItem(true)
        ;
        searchResult = client.collections(collection).documents().search(searchParameters);
        hits = searchResult.getHits();
        assertEquals(hits.get(0).getDocument().get("id"), "doc2");
        assertEquals(hits.get(1).getDocument().get("id"), "doc1");


        searchParameters = new SearchParameters()
                .q("foo same").addQueryByItem("fieldA,fieldB").addQueryByWeightsItem("1,125").addPrefixItem(true)
        ;
        searchResult = client.collections(collection).documents().search(searchParameters);
        hits = searchResult.getHits();
        assertEquals(hits.get(0).getDocument().get("id"), "doc2");
        assertEquals(hits.get(1).getDocument().get("id"), "doc1");

        searchParameters = new SearchParameters()
                .q("foo same").addQueryByItem("fieldA,fieldB").addQueryByWeightsItem("1,128").addPrefixItem(true)
        ;
        searchResult = client.collections(collection).documents().search(searchParameters);
        hits = searchResult.getHits();
        //==============================================================================================
        // The following asserts fail
        //==============================================================================================
        assertEquals(hits.get(0).getDocument().get("id"), "doc2");
        assertEquals(hits.get(1).getDocument().get("id"), "doc1");
`

**Comment by kishorenc** [comment]
@superkelvint 

> I found something else that might indicate you're using a signed byte to represent query_by_weight values because for weights > 127, they seem to be ignored or inverted. 

Good catch. The various components of match scoring indeed use a byte so we should certainly document that. Please create a separate issue for tracking this 

>  I just tried against 0.22.rcs37 and the test still fails. I can provide the unit test in a different language if you can't run the Java class easily.

That will be great. Can you please provide a curl snippet or Python client example? I will then be able to debug it locally and get back to you.

**Comment by kishorenc** [comment]
@superkelvint 

Just got around to looking into this in more detail. 

I've fixed the issue with weights > 127 in a recent commit. As for the other weight related issue:

> Adding a query_by_weight parameter of "1,1" for a 2-field search causes ranking to change which appears to violate the documentation.

When you query fields _without weights_ Typesense treats fields higher in the list of fields to be higher weighted than fields lower in the list -- it does not give the fields equal weights. So the following two queries **won't** have the same weights:

```
// Query 1 (no explicit weights given)
// Typesense will implicitly assign higher weight to fieldA than fieldB
SearchParameters searchParameters = new SearchParameters()
            .q("foo same").addQueryByItem("fieldA,fieldB").addPrefixItem(true);
SearchResult searchResult = client.collections(collection).documents().search(searchParameters);

// Query 2 (explicit equal weights given)
searchParameters = new SearchParameters()
            .q("foo same").addQueryByItem("fieldA,fieldB").addQueryByWeightsItem("1,1").addPrefixItem(true);
searchResult = client.collections(collection).documents().search(searchParameters);
```

I noticed that we don't talk about this default weighting behavior in the documentation -- I will fix that shortly. Please let me know if I have missed something else here, thanks.
